src/measure_theory/covering/density_theorem.lean

```diff
@@ -0,0 +1,148 @@
+/-
+Copyright (c) 2022 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import measure_theory.covering.vitali
+import measure_theory.covering.differentiation
+import analysis.special_functions.log.base
+
+/-!
+# Doubling measures and Lebesgue's density theorem
+
+A doubling measure `μ` on a metric space is a measure for which there exists a constant `C` such
+that for all sufficiently small radii `ε`, and for any centre, the measure of a ball of radius
+`2 * ε` is bounded by `C` times the measure of the concentric ball of radius `ε`.
+
+Lebesgue's density theorem states that given a set `S` in a proper metric space with locally-finite
+doubling measure `μ` then for almost all points `x` in `S`, for any sequence of closed balls
+`B₀, B₁, B₂, ...` containing `x`, the limit `μ (S ∩ Bⱼ) / μ (Bⱼ) → 1` as `j → ∞`.
+
+In this file we combine general results about existence of Vitali families for doubling measures
+with results about differentiation along a Vitali family to obtain an explicit form of Lebesgue's
+density theorem.
+
+## Main results
+
+  * `is_doubling_measure`: the definition of a doubling measure (as a typeclass).
+  * `is_doubling_measure.doubling_constant`: a function yielding the doubling constant `C` appearing
+  in the definition of a doubling measure.
+  * `is_doubling_measure.ae_tendsto_measure_inter_div`: a version of Lebesgue's density theorem for
+  sequences of balls converging on a point but whose centres are not required to be fixed.
+
+-/
+
+noncomputable theory
+
+open set filter metric measure_theory
+open_locale nnreal topological_space
+
+/-- A measure `μ` is said to be a doubling measure if there exists a constant `C` such that for
+all sufficiently small radii `ε`, and for any centre, the measure of a ball of radius `2 * ε` is
+bounded by `C` times the measure of the concentric ball of radius `ε`.
+
+Note: it is important that this definition makes a demand only for sufficiently small `ε`. For
+example we want hyperbolic space to carry the instance `is_doubling_measure volume` but volumes grow
+exponentially in hyperbolic space. To be really explicit, consider the hyperbolic plane of
+curvature -1, the area of a disc of radius `ε` is `A(ε) = 2π(cosh(ε) - 1)` so `A(2ε)/A(ε) ~ exp(ε)`.
+-/
+class is_doubling_measure {α : Type*} [metric_space α] [measurable_space α] (μ : measure α) :=
+(exists_measure_closed_ball_le_mul [] :
+  ∃ (C : ℝ≥0), ∀ᶠ ε in 𝓝[>] 0, ∀ x, μ (closed_ball x (2 * ε)) ≤ C * μ (closed_ball x ε))
+
+namespace is_doubling_measure
+
+variables {α : Type*} [metric_space α] [measurable_space α] (μ : measure α) [is_doubling_measure μ]
+
+/-- A doubling constant for a doubling measure. -/
+def doubling_constant : ℝ≥0 := classical.some $ exists_measure_closed_ball_le_mul μ
+
+lemma exists_measure_closed_ball_le_mul' :
+  ∀ᶠ ε in 𝓝[>] 0, ∀ x, μ (closed_ball x (2 * ε)) ≤ (doubling_constant μ) * μ (closed_ball x ε) :=
```

Comment:
This mistakenly looks like function application.
```suggestion
  ∀ᶠ ε in 𝓝[>] 0, ∀ x, μ (closed_ball x (2 * ε)) ≤ doubling_constant μ * μ (closed_ball x ε) :=
```

Reply:
Thanks; redundant brackets gone.